ava.1
.TH AVA 1 ava\-VERSION
.SH NAME
ava \- assitance virtual answer
.SH SYNOPSIS
.DS
\fB ava [OPTIONS]...\ [QUERY]...\fP
\fB ava [QUERY]...\ [OPTIONS]...\fP
.SH DESCRIPTION
Ava is an innovative virtual assistant designed to help you find information
online quickly and easily. With its intuitive interface and powerful search
capabilities, Ava makes it easy to get the answers you need without the hassle
of navigating through Google search engines. Whether you need to solve a math
equalition, look up the lyrics to your favorite song, or translate a piece of
text, Ava has got you covered.
.P
One of the key features of Ava is its flexibility. Unlike other virtual
assistant that are limited in their scope, Ava supports a wide range of search
types, making it a valuable tool for anyone looking for information on the web.
From basic arithmatic to complex mathematical expressions, Ava's math search
type can handle it all. And with its lyrics search type, you can quickly and
easily find can words to any song, no matter how obscure. The translation search
type is also incredibly useful for anyone who needs to communicate in different
languages.
.P
Ava's user-friendly interface is another major advantage. With its simple,
streamlined design, Ava is easy to use, even for those who are not tech-savvy.
The interface is intuitive and responsive, making it a pleasure to work with.
And with its advanced search capabilities, Ava can provide results that are more
accurate and relevant than those offered by Google search engines and tools.
.P
Overall, Ava is a powerful and versatile tool that can save you time and
frustration when looking for information online. With its innovative features
and user-friendly interface, Ava is a must-have tool for anyone who needs to
find information quickly and easily.
.P
So why wait? Download Ava and start
enjoying the benefits of this amazing virtual assistant!
.SH OPTIONS
.SS Global
.TP
.B \-h, --help
Show this help message and exit.
.TP
.B \-v, --version
Print ava version info and exit.
.TP
.B \-r, --raw
Raw search results. (no pretty output, no colors)
.TP
.B \-q, --quiet
Only output search results. (silences "Did you mena?", greeting, usage)
.TP
.B \-a, --all
Print all valid answers.
.TP
.B \-u, --plus_urls
Prints out the top handful or URLs for your search query. (this is automatically
printed out if ava can't find you an answer)
.SS Developer Flags
.TP
.B \-d, --debug
Prints debug info along with results
.TP
.B \-s, --save
saves HTML for this query to $XDG_CACHE_HOME/ava/ava-[date].html
.SH USAGE
Ava is designed to be easy to use and intuitive. To use Ava, simply run the
`ava` command followed by the options flags you want to perform and any relevant
query parameters. For example:
.P
.B $ ava "2+2"
.P
This command will search for the solution to the math equation "2+2" and display
the result in the terminal.
.SH ISSUES
In using Ava, there are several issues that may arise. One common issue is the need to execute JavaScript correctly, ensuring that there are no syntax errors or function call issues. Regular maintenance is also essential, including bug fixes, security updates, and functional improvements. To address the problem of unappealing and messy output, you can use appropriate CSS styles to organize the display of HTML elements and consider using front-end frameworks that provide ready-made components and styles. Additionally, it is crucial to establish a well-organized project structure by separating files based on their functions and responsibilities and providing descriptive file and directory names.
.P
Furthermore, there are several other issues that may occur, such as the need to optimize code. It is important to identify slow or inefficient sections of code in your project and make necessary improvements, such as fixing inefficient algorithms or excessive resource usage. Employing optimization techniques like caching, code merging and compression, lazy loading, or asynchronous programming can also be beneficial. Lastly, addressing the problem of poorly named variables is crucial. By providing clear and descriptive names for variables and following consistent naming conventions, you can enhance code readability. Additionally, adding clear comments can help explain the usage of complex variables.
.SH BUGS
Send all bug report to fajar.nur.h@pm.me or https://github.com/0xFA99/Ava/issues



ava.c
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>
#include <libxml2/libxml/HTMLparser.h>
#include <curl/curl.h>
#include <getopt.h>

#include "ava.h"

#include "config.def.h"
#include "utils.h"
#include "answer.h"

unsigned int
handleMemoryCallback(void *content, size_t size, size_t nmemb, void *userp)
{
	struct HttpResponse *response = (struct HttpResponse *) userp;
	size_t realsize = size * nmemb;
	char *new_data = realloc(response->responseData, response->responseSize + realsize + 1);
	if (new_data == NULL) {
		fprintf(stderr, "Out of Memory!\n");
		return 0;
	}

	response->responseData = new_data;
	memcpy(&(response->responseData[response->responseSize]), content, realsize);
	response->responseSize += realsize;
	response->responseData[response->responseSize] = '\0';

	return realsize;
}

char*
createURL(const char *lang, const char *text)
{
	unsigned int lang_len = strlen("hl=") + strlen(lang);
	unsigned int text_len = strlen("q=") + strlen(text);
	char *language	= malloc(lang_len + 1);
	char *query		= malloc(text_len + 1);

	snprintf(language, lang_len + 1, "%s%s", "hl=", lang);
	snprintf(query, text_len + 1, "%s%s", "q=", text);

	CURLU *newURL = curl_url();
	curl_url_set(newURL, CURLUPART_SCHEME, "https", 0);
	curl_url_set(newURL, CURLUPART_URL, defaultSearchEngineUrl, 0);
	curl_url_set(newURL, CURLUPART_PATH, "search", 0);
	curl_url_set(newURL, CURLUPART_QUERY, language, CURLU_APPENDQUERY);
	curl_url_set(newURL, CURLUPART_QUERY, query, CURLU_APPENDQUERY | CURLU_URLENCODE);

	char *result = NULL;
	curl_url_get(newURL, CURLUPART_URL, &result, 0);

	char *dump_result = strdup(result);

	free(language);
	free(query);

	curl_url_cleanup(newURL);

	return dump_result;
}

char*
sendHttpsRequest(const char *query)
{
	char *defaultURL = createURL("en_US", query);
	CURL *curl;
	CURLcode res;
	struct HttpResponse response;

	response.responseData = malloc(1024 * 1024);
	response.responseSize = 0;

	curl = curl_easy_init();
	if (curl) {
		curl_easy_setopt(curl, CURLOPT_URL, defaultURL);
		curl_easy_setopt(curl, CURLOPT_USERAGENT, defaultUserAgent);
		curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, handleMemoryCallback);
		curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);

		res = curl_easy_perform(curl);

		if (res != CURLE_OK) {
			fprintf(stderr, "curl_easy_perform, failed: %s\n", curl_easy_strerror(res));
			free(response.responseData);
			response.responseData = NULL;
		}

		curl_easy_cleanup(curl);
	}

	free(defaultURL);

	return response.responseData;
}

int
main(int argc, char *argv[])
{
	char *query = NULL;

	struct SearchFlags SFlags = { 
		.raw		= false,
		.quiet		= false,
		.all		= false,
		.debug		= false,
		.save_html	= false,
		.plus_urls	= false
	};
	struct option long_opts[] = {
			{ "raw",        no_argument, NULL, 'r' },
			{ "version",    no_argument, NULL, 'v' },
			{ "help",       no_argument, NULL, 'h' },
			{ "quiet",      no_argument, NULL, 'q' },
			{ "all",        no_argument, NULL, 'a' },
			{ "debug",      no_argument, NULL, 'd' },
			{ "save",       no_argument, NULL, 's' },
			{ "plus_urls",  no_argument, NULL, 'u' }
	};

	int opt;
	while ((opt = getopt_long(argc, argv, "rvhqadsu", long_opts, NULL)) != -1) {
		switch (opt) {
			case 'r':
				SFlags.raw = true; break;
			case 'v':
				exitWithError("AVA Version: %s\n", VERSION); return 1;
			case 'h':
				exitWithError(GREEN "Usage: " RESET "ava " YELLOW "[options] " RESET MAGENTA "\"query\"\n" RESET);
				return 1;
			case 'q':
				SFlags.quiet = true; break;
			case 'a':
				SFlags.all = true; break;
			case 'd':
				SFlags.debug = true; break;
			case 's':
				SFlags.save_html = true; break;
			case 'u':
				SFlags.plus_urls = true; break;
			case '?':
				exitWithError("Invalid option\n"); break;
			default:
				exitWithError("Unknown option\n"); break;
		}
	}

	if (optind < argc) {
		query = argv[optind];
		if (strlen(query) == 0) {
			exitWithError("Query: (empty)\n");
			return 1;
		}
	} else {
		exitWithError(GREEN "Usage: " RESET "ava " YELLOW "[options] " MAGENTA "\"query\"\n" RESET);
		return 1;
	}

	char *response = NULL;

	if (SFlags.debug) {
		struct timespec start_time, end_time;
		double response_time, parsing_time;

		getCurrentTime(&start_time);
		response = sendHttpsRequest(query);
		getCurrentTime(&end_time);

		response_time = (double)(end_time.tv_sec - start_time.tv_sec) +
						(double)(end_time.tv_nsec - start_time.tv_nsec) / 1e9;

		if (SFlags.save_html)
			saveResponseToFile(response);

		getCurrentTime(&start_time);
		parseHTML(SFlags, response);
		getCurrentTime(&end_time);

		parsing_time = (double)(end_time.tv_sec - start_time.tv_sec) +
					   (double)(end_time.tv_nsec - start_time.tv_nsec) / 1e9;

		printf("Response Time: %.3fs\n", response_time);
		printf("Parsing Time: %.3fs\n", parsing_time);
	} else {
		response = sendHttpsRequest(query);

		if (SFlags.save_html)
			saveResponseToFile(response);

		parseHTML(SFlags, response);
	}

	free(response);

    return 0;
}
